### PR TITLE
Introduce operator fixity configuration

### LINKED ISSUE #845: Implement support for operator fixity configuration file
*This issue describes the next step regarding the operator chains overhaul, it is based on the work in #830.*

We need to introduce the ability to overwrite fixity info of specific operators selectively. The override should be added to the `Config` type and can have exactly the same type as normal fixity map. There should be also a mechanism to read `.ormolu` (suggested name for configuration files) automatically if it exists, otherwise do nothing leaving the overriding fixity map empty.

I propose we use the same familiar syntax in `.ormolu` for fixity information, e.g.:

```haskell
infixl 9  !!
infix  4 `elem`, `notElem`
```

This has the advantage that it is immediately editable by Haskellers without a need to study a new syntax.

Further, the new override mechanism should be used in the test suite instead of our current practice of choosing a few packages that have the right operators in order to get the "right" results in the tests. This should make the test suite a little bit more robust.

### PR DESCRIPTION
**Pull request by mrkkrp**
Close #845.

### CODE REVIEW

File: src/Ormolu/Utils/Fixity.hs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+-- | Attempt to locate and parse a @.ormolu@ file. If it does not exist,
+-- empty fixity map is returned. This function maintains a cache of fixity
+-- overrides where cabal file paths act as keys.
+getFixityOverridesForSourceFile ::
+  MonadIO m =>
+  -- | 'CabalInfo' already obtained for this source file
+  CabalInfo ->
+  m FixityMap
+getFixityOverridesForSourceFile CabalInfo {..} = liftIO $ do
+  cacheRef <- newIORef Map.empty
```
Comment (amesgen) [inline src/Ormolu/Utils/Fixity.hs:33]: This `cacheRef` is only used locally in this function (in particular, the subsequent `readIORef cacheRef` will always result in `Map.empty`). This should probably passed in here as an argument in some way?

File: src/Ormolu/Utils/Fixity.hs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+          exists <- doesFileExist dotOrmolu
+          if exists
+            then do
+              dotOrmoluRelative <- makeRelativeToCurrentDirectory dotOrmolu
+              contents <- T.readFile dotOrmolu
+              case parseFixityMap dotOrmoluRelative contents of
+                Left errorBundle ->
+                  throwIO (OrmoluFixityOverridesParseError errorBundle)
+                Right x -> do
+                  modifyIORef cacheRef (Map.insert cabalPath x)
```
Comment (amesgen) [inline src/Ormolu/Utils/Fixity.hs:50]:
```suggestion
                  modifyIORef' cacheRef (Map.insert cabalPath x)
```

File: src/Ormolu/Utils/Fixity.hs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    Just cabalPath -> do
+      cache <- readIORef cacheRef
+      case Map.lookup cabalPath cache of
+        Nothing -> do
+          let dotOrmolu = replaceFileName cabalPath ".ormolu"
+          exists <- doesFileExist dotOrmolu
+          if exists
+            then do
+              dotOrmoluRelative <- makeRelativeToCurrentDirectory dotOrmolu
+              contents <- T.readFile dotOrmolu
```
Comment (amesgen) [inline src/Ormolu/Utils/Fixity.hs:45]:
```suggestion
              contents <- readFileUtf8 dotOrmolu
```
and importing `Ormolu.Utils.IO (readFileUtf8)`.

**Review by amesgen — CHANGES_REQUESTED** [review]
Looks great; LGTM from the comments in `Ormolu.Utils.Fixity` :+1:

Also, the test suite slowed down a bit, on my machine from 1.0s to 2.3s, probably due the `testsuiteFixities` being inserted into the huge fixity map repeatedly? (but I haven't looked into it too deply)

File: src/Ormolu/Utils/Cabal.hs [hunk 4]
```diff
@@ -40,11 +39,13 @@ import System.IO.Error (isDoesNotExistError)
...
-    ciPackageName :: Maybe String,
+    ciPackageName :: !(Maybe String),
     -- | Extension and language settings in the form of 'DynOption's
-    ciDynOpts :: [DynOption],
+    ciDynOpts :: ![DynOption],
     -- | Direct dependencies
-    ciDependencies :: Set String
+    ciDependencies :: !(Set String),
+    -- | Absolute path to the cabal file, if it was found
+    ciCabalFilePath :: !(Maybe FilePath)
```
[the remark this answers is not in the document]
Comment (mrkkrp) [reply src/Ormolu/Utils/Cabal.hs:48]: This is not for strictness, this is to make compilation fail when this field is not initialized.

File: CHANGELOG.md [hunk 5]
```diff
@@ -23,12 +14,28 @@
...
   * The default heuristic algorithm will still try to guess the right
     fixities and precedence.
-  This resolves the following issues: [Issue
+  * Fixity overrides can be provided by the user in the familiar Haskell
+    notation (e.g. `infixr 9 .`, one declaration per line). They are loaded
+    by default from the `.ormolu` file that is expected to be in the same
+    directory as the `.cabal` file of a given source file. However, if
+    `--no-cabal` is supplied `.ormolu` file will not be looked for either.
+    Fixity declarations can be also provided by using the `-f / --fixity`
+    command line option, which see.
```
[the remark this answers is not in the document]
Comment (mrkkrp) [reply CHANGELOG.md:23]: I think it is complete. You can say, e.g.

> The first part of this quotation is based upon the Emerald Tablet of Hermes (which see).

File: src/Ormolu/Utils/Cabal.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (amesgen) [reply src/Ormolu/Utils/Cabal.hs:48]: Hmm, if that is the reason, then I would suggest to enable the `-Werror` (e.g. via the `dev` flag) on CI, such that missing initialized fields are caught even for lazy records (except for the `GHC.DynFlags` module, where `-Wmissing-fields` is already intentionally disabled). WDYT?

File: CHANGELOG.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (amesgen) [reply CHANGELOG.md:23]: Ah thanks, didn't know that one; this seems to be the literal translation of a latin phrase, ["quod vide"](https://en.wikipedia.org/wiki/List_of_Latin_phrases_(full)#quod_vide=).

File: src/Ormolu/Utils/Cabal.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mrkkrp) [reply src/Ormolu/Utils/Cabal.hs:48]: We already compile with `+dev` and it already catches this. I was mislead by Haskell LSP server which showed this error as a warning.

File: src/Ormolu/Utils/Fixity.hs (hunk 2 above)
Comment (mrkkrp) [reply src/Ormolu/Utils/Fixity.hs]: I'm not sure we are gaining here anything from strictness, but it doesn't hurt either!

**Review by amesgen — APPROVED** [review]

File: src/Ormolu/Utils/Fixity.hs (hunk 2 above)
Comment (amesgen) [reply src/Ormolu/Utils/Fixity.hs]: In contrived cases (lots of files, each with their own `.ormolu` file, so cache entries are never read), there could be a thunk buildup in `cacheRef`.

**Pull request merged**